File: markdown.go
package main

import (
	"bytes"
	"fmt"
	"html/template"
	"io/ioutil"
	"net/http"
	"regexp"
	"strings"
)

const port = 8365

const monokaiTemplate = `<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8">
<title>Markdown Viewer - Monokai</title>
<style>
body {
	background-color: #272822;
	color: #f8f8f2;
	font-family: 'Consolas', 'Monaco', monospace;
	line-height: 1.5;
	padding: 20px;
	max-width: 900px;
	margin: 0 auto;
}
a {
	color: #66d9ef;
	text-decoration: none;
}
a:hover {
	text-decoration: underline;
}
h1, h2, h3, h4, h5, h6 {
	color: #f92672;
	margin-top: 24px;
	margin-bottom: 16px;
}
code {
	background-color: #3e3d32;
	color: #f8f8f2;
	padding: 0.2em 0.4em;
	border-radius: 3px;
	font-family: 'Consolas', 'Monaco', monospace;
}
pre {
	background-color: #3e3d32;
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}
pre code {
	padding: 0;
	background-color: transparent;
}
blockquote {
	border-left: 4px solid #66d9ef;
	padding-left: 16px;
	color: #75715e;
	margin-left: 0;
}
hr {
	border: 0;
	height: 1px;
	background-color: #75715e;
	margin: 24px 0;
}
table {
	border-collapse: collapse;
	width: 100%;
	margin-bottom: 16px;
}
th, td {
	border: 1px solid #75715e;
	padding: 6px 13px;
}
th {
	background-color: #3e3d32;
}
</style>
</head>
<body>
{{.}}
</body>
</html>`

func main() {
	http.HandleFunc("/", handleRequest)
	fmt.Printf("Сервер запущен на http://localhost:%d\n", port)
	fmt.Printf("Пример запроса: http://localhost:%d?file=README.md\n", port)
	err := http.ListenAndServe(fmt.Sprintf(":%d", port), nil)
	if err != nil {
		fmt.Printf("Ошибка запуска сервера: %v\n", err)
	}
}

func handleRequest(w http.ResponseWriter, r *http.Request) {
	filePath := r.URL.Query().Get("file")
	if filePath == "" {
		http.Error(w, "Параметр 'file' не указан", http.StatusBadRequest)
		return
	}

	// Безопасная проверка пути
	if strings.Contains(filePath, "..") {
		http.Error(w, "Недопустимый путь к файлу", http.StatusBadRequest)
		return
	}

	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		http.Error(w, fmt.Sprintf("Ошибка чтения файла: %v", err), http.StatusInternalServerError)
		return
	}

	// Конвертируем Markdown в HTML (упрощенная реализация)
	htmlContent := markdownToHTML(content)

	// Создаем шаблон с темой Monokai
	tmpl, err := template.New("monokai").Parse(monokaiTemplate)
	if err != nil {
		http.Error(w, fmt.Sprintf("Ошибка создания шаблона: %v", err), http.StatusInternalServerError)
		return
	}

	// Вставляем сгенерированный HTML в шаблон
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	err = tmpl.Execute(w, template.HTML(htmlContent))
	if err != nil {
		http.Error(w, fmt.Sprintf("Ошибка рендеринга шаблона: %v", err), http.StatusInternalServerError)
		return
	}
}

// Упрощенный парсер Markdown -> HTML
func markdownToHTML(md []byte) string {
	// Сначала обрабатываем блоки кода (```...```), чтобы их содержимое не парсилось дальше
	md = regexp.MustCompile("(?s)```[a-z]*\\n(.+?)```").ReplaceAll(md, []byte("<pre><code>$1</code></pre>"))
	md = regexp.MustCompile("(?s)~~~[a-z]*\\n(.+?)~~~").ReplaceAll(md, []byte("<pre><code>$1</code></pre>"))

	// Затем обрабатываем инлайновый код (`...`), пропуская уже обработанные блоки
	md = regexp.MustCompile("(?s)(<pre>.*?</pre>)|`([^`]+)`").ReplaceAllFunc(md, func(m []byte) []byte {
		if bytes.HasPrefix(m, []byte("<pre>")) {
			return m // уже обработанный блок кода
		}
		return []byte("<code>" + string(m[1:len(m)-1]) + "</code>")
	})

	// Обрабатываем заголовки
	md = regexp.MustCompile(`(?m)^#{1,3}\s(.+)$`).ReplaceAllFunc(md, func(m []byte) []byte {
		level := bytes.Count(m, []byte("#"))
		content := bytes.TrimSpace(m[level:])
		return []byte(fmt.Sprintf("<h%d>%s</h%d>", level, content, level))
	})

	// Обрабатываем ссылки
	md = regexp.MustCompile(`\[([^\]]+)\]\(([^)]+)\)`).ReplaceAll(md, []byte(`<a href="$2">$1</a>`))

	// Обрабатываем жирный текст (два варианта: ** и __)
	md = regexp.MustCompile(`\*\*(.+?)\*\*`).ReplaceAll(md, []byte(`<strong>$1</strong>`))
	md = regexp.MustCompile(`__(.+?)__`).ReplaceAll(md, []byte(`<strong>$1</strong>`))

	// Обрабатываем курсив (два варианта: * и _)
	md = regexp.MustCompile(`\*(.+?)\*`).ReplaceAll(md, []byte(`<em>$1</em>`))
	md = regexp.MustCompile(`_(.+?)_`).ReplaceAll(md, []byte(`<em>$1</em>`))

	// Обрабатываем цитаты
	md = regexp.MustCompile(`(?m)^>\s(.+)$`).ReplaceAll(md, []byte(`<blockquote>$1</blockquote>`))

	// Обрабатываем горизонтальные линии
	md = regexp.MustCompile(`(?m)^---$`).ReplaceAll(md, []byte(`<hr/>`))

	// Обрабатываем списки
	md = regexp.MustCompile(`(?m)^[*+-]\s(.+)$`).ReplaceAll(md, []byte(`<li>$1</li>`))
	md = regexp.MustCompile(`(?s)(<li>.+?</li>\s*)+`).ReplaceAllFunc(md, func(m []byte) []byte {
		return []byte("<ul>" + string(m) + "</ul>")
	})

	// Заменяем переносы строк на <br>, но не внутри <pre> блоков
	md = regexp.MustCompile(`(?s)(<pre>.*?</pre>)|(\n)`).ReplaceAllFunc(md, func(m []byte) []byte {
		if len(m) == 1 && m[0] == '\n' {
			return []byte("<br>")
		}
		return m
	})

	return string(md)
}

File: libs/markdown/markdown.php
#!/bin/php
<?php
include "Server.php";
include "Request.php";
include "Response.php";
include "ParseMarkdown.php";
include "MD.php";

$server = new Server("127.0.0.1", 8365);

$server->listen(function($rq) {
	try {
		if (count($rq->parameters) < 1 || (!isset($rq->parameters['file']))) {
			return new Response("", 200);
		}
	$file = $rq->parameters['file'];
	$sample = file_get_contents($file);


	# init Markdown object
	$mkd = Markdown::new();

	# set markdown data to convert
	$mkd -> setContent( $sample );

	# convert data to markdown
	$MD = $mkd -> toHtml();
$html = <<<HTML
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>MD</title>
    <style>
        :root {
            --bg-color: #272822;
            --text-color: #F8F8F2;
            --primary: #F92672;
            --secondary: #A6E22E;
            --tertiary: #66D9EF;
            --quaternary: #FD971F;
            --code-bg: #3E3D32;
            --code-border: #75715E;
            --blockquote-bg: #3E3D32;
            --blockquote-border: #75715E;
            --table-border: #75715E;
            --table-header-bg: #3E3D32;
            --table-row-alt-bg: #2D2E27;
            --link-color: #66D9EF;
            --link-hover: #A6E22E;
        }
        
        body {
            background-color: var(--bg-color);
            color: var(--text-color);
            font-family: 'Consolas', 'Monaco', 'Courier New', monospace;
            line-height: 1.6;
            margin: 0;
            padding: 20px;
            max-width: 900px;
            margin: 0 auto;
        }
        
        a {
            color: var(--link-color);
            text-decoration: none;
        }
        
        a:hover {
            color: var(--link-hover);
            text-decoration: underline;
        }
        
        h1, h2, h3, h4, h5, h6 {
            color: var(--primary);
            margin-top: 24px;
            margin-bottom: 16px;
        }
        
        h1 { border-bottom: 1px solid var(--code-border); }
        h2 { border-bottom: 1px solid var(--code-border); }
        
        code {
            background-color: var(--code-bg);
            color: var(--tertiary);
            padding: 2px 4px;
            border-radius: 3px;
            font-family: monospace;
        }
        
        pre {
            background-color: var(--code-bg);
            border: 1px solid var(--code-border);
            border-radius: 4px;
            padding: 12px;
            overflow: auto;
        }
        
        pre code {
            background-color: transparent;
            color: var(--text-color);
            padding: 0;
        }
        
        blockquote {
            background-color: var(--blockquote-bg);
            border-left: 4px solid var(--blockquote-border);
            margin: 0 0 16px 0;
            padding: 0 16px;
            color: #ddd;
        }
        
        table {
            border-collapse: collapse;
            width: 100%;
            margin-bottom: 16px;
        }
        
        th, td {
            border: 1px solid var(--table-border);
            padding: 8px 12px;
        }
        
        th {
            background-color: var(--table-header-bg);
            color: var(--secondary);
        }
        
        tr:nth-child(even) {
            background-color: var(--table-row-alt-bg);
        }
        
        img {
            max-width: 100%;
            height: auto;
        }
        
        hr {
            border: 0;
            height: 1px;
            background-color: var(--code-border);
            margin: 24px 0;
        }
        
        .highlight {
            background-color: var(--quaternary);
            color: var(--bg-color);
            padding: 2px 4px;
            border-radius: 3px;
        }
    </style>
</head>
<body>
    $MD
</body>
</html>
HTML;
	$resp = new Response($html, 200);
	$resp->header("Content-Type", "text/html");
	return $resp;
	} catch (Exception $e) {}
}, function(){});
